design/StableMemory.md

```diff
@@ -249,19 +259,40 @@ fun restore() : value option =
       else None
     else
       size := 0;
-      let len = mem[0,..,3] in
+      let len = marker in
       assert (0 < len <= ic0.stable_size() * pagesize)
       let v = deserialise(len, 4) in
       mem[0, len + 1] := 0 // clear memory
       Some v
-
-(* Note we explicitly clear memory used by stable variables so StableMem doesn't need to clear memory
-   when grabbing logical pages from already existing physical ones *).
 ```
 
+We explicitly clear memory used by stable variables so StableMem
+doesn't need to clear memory when grabbing logical pages from already
+existing physical ones.
 
 
-Note that we still need to do some work during updgrade and postupgrade, but if stable variables and user-defined pre/post upgrade hooks are
-avoided, then the work is minimal and highly unlikely to exhaust cycle budget.
+NOTE: We still need to do some work during updgrade and postupgrade,
+but if stable variables and user-defined pre/post upgrade hooks are
+avoided, then the work is minimal and highly unlikely to exhaust cycle
+budget.
 
+REMARK:
 
+* An actor that has no stable variables and allocates no StableMem
+  should requize no physical stable memory
+
+* An actor that has n > 0 pages of StableMem will (unfortunately)
+  require at least n+1 pages of physical memory since we need at least
+  one extra bit to encode the presence or absence of stable variables
+  (there is no other preserved state that could record this bit).
+
+FURTHER CONSIDERATIONS: It would be nice if there was some way to
+allow a Motoko actor (perhaps intended to upgrade from a foreign
+canister with foreign stable memory format) to have unadulterated,
+full-speed access to stable memory, without the protocol required by
+the possibility of stable variables.  Perhaps we could special case
+programs that have *no* stable variables to support this raw
+semantics.  I.e. if the program declares no stable variables, we
+install and upgrade by simply setting `(StableMemory.)size :=
+ic0.stable_size()`, never consulting or altering physical memory and
+(ideally) omitting the additional bounds checking.
```

Comment:
Yeah, I was thinking about this as well. But it seems like we always have to take into account the possibility that there were stable variables before the upgrade, written by the prior code. Or that at some point in time there was an even older version that grew the physical size of stable memory past (the current) StableMemory.size. Dammit.

Something else I wondered was whether we could somehow make it so that StableMem is always located at the _end_ of stable memory -- then instead of custom bounds check we'd only need to add an offset, which should be cheaper. But that would require shifting all of StableMem each time stable vars are backed up whose value takes more space than what was previously reserved. That would be pretty costly. OTOH, that cost would never come up if you didn't use both kinds of stable data at the same time. But still probably not worth it.

Reply:
> Yeah, I was thinking about this as well. But it seems like we always have to take into account the possibility that there were stable variables before the upgrade, written by the prior code.

But can't we just ignore them, and let the (sv-less) code decide what to do with the populated memory?

> Or that at some point in time there was an even older version that grew the physical size of stable memory past (the current) StableMemory.size. Dammit.

Is that a problem? At that point we just set StableMemory.size to ic0.stable_size() and the actor can see the full memory and do as they see fit? If they want to keep a finger on StableMemory, they can use a dedicated address and upgrade hook to save it. If we wanted to restore it,  we'd need to extend the interface.
I think you could even use this technique to migrate code from pure stable variables to pure StableMemory, provided we give people access to Candid(ish) deserialization.

> Something else I wondered was whether we could somehow make it so that StableMem is always located at the _end_ of stable memory -- then instead of custom bounds check we'd only need to add an offset, which should be cheaper. But that would require shifting all of StableMem each time stable vars are backed up whose value takes more space than what was previously reserved. That would be pretty costly. OTOH, that cost would never come up if you didn't use both kinds of stable data at the same time. But still probably not worth it.

Reply:
It would be fishy to arbitrarily change StableMem.size behind the app's back. It might have certain assumptions attached to the size it knows, i.e., that the page number is a multiple of something or that each page is initialised in a certain way.

So I think you can perhaps simplify some logic if they _are_ already equal. But you cannot _make_ them equal.